File: adjacency_matrix/go/adjacency_matrix.go
package adjacency_matrix

type Graph struct {
	VertexCount int
	Directed    bool
	Edges       []Edge
}

type Edge struct {
	X int
	Y int
}

type AdjacencyMatrix [][]bool

type VertexCallback func(x int)

func NewAdjacencyMatrix(graph Graph) *AdjacencyMatrix {
	m := initAdjacencyMatrix(graph.VertexCount)
	for _, edge := range graph.Edges {
		m.insertEdge(edge.X, edge.Y, graph.Directed)
	}
	return m
}

func initAdjacencyMatrix(vertexCount int) *AdjacencyMatrix {
	m := make(AdjacencyMatrix, vertexCount)
	for i := range m {
		m[i] = make([]bool, vertexCount)
	}
	return &m
}

func (m *AdjacencyMatrix) insertEdge(x, y int, directed bool) {
	(*m)[x][y] = true
	if !directed {
		m.insertEdge(y, x, true)
	}
}

func (m *AdjacencyMatrix) BreadthFirstSearch(start int, callback VertexCallback) {
	discovered := make([]bool, len(*m))
	queue := []int{}

	discovered[start] = true
	callback(start)
	queue = append(queue, start)

	for len(queue) > 0 {
		x := queue[0]
		queue = queue[1:]
		for y, connected := range (*m)[x] {
			if !connected || discovered[y] {
				continue
			}
			discovered[y] = true
			callback(y)
			queue = append(queue, y)
		}
	}
}

func (m *AdjacencyMatrix) DepthFirstSearch(start int, preOrderCallback VertexCallback) {
	discovered := make([]bool, len(*m))
	m.dfs(start, preOrderCallback, discovered)
}

func (m *AdjacencyMatrix) dfs(x int, preOrderCallback VertexCallback, discovered []bool) {
	if discovered[x] {
		return
	}
	discovered[x] = true
	preOrderCallback(x)
	for y, connected := range (*m)[x] {
		if connected {
			m.dfs(y, preOrderCallback, discovered)
		}
	}
}

// func (m *AdjacencyMatrix) String() string {
// 	var b bytes.Buffer
// 	b.WriteString("\n")
// 	for i, row := range *m {
// 		b.WriteString(fmt.Sprintf("%d: %v\n", i, row))
// 	}
// 	return b.String()
// }

File: bipartite_matching/typescript/adjacency_matrix.ts
export interface Graph {
    directed: boolean;
    vertexCount: number;
    edges: Edge[];
}

export type Edge = number[];

export interface BipartiteGraph extends Graph {
    flags: boolean[];
}

type EdgeCallback = (edge: Edge) => boolean;

export class AdjacencyMatrix {
    a: boolean[][];

    constructor(graph: Graph) {
        this.a = constructAdjacencyMatrix(graph);
    }

    breadthFirstSearch(start: number, callback: EdgeCallback) {
        const visited = new Array(this.a.length);
        visited[start] = true;

        const queue = [start];

        while (queue.length) {
            const x = queue.shift();

            for (let y = 0; y < this.a.length; y++) {
                if (visited[y] || !this.a[x][y]) {
                    continue;
                }
                visited[y] = true;
                queue.push(y);
                // The edge callback can return true to abort the current graph traversal.
                if (callback([x, y])) {
                    return;
                }
            }
        }
    }
}

function constructAdjacencyMatrix(graph: Graph): boolean[][] {
    const a = Array.from(
        {length: graph.vertexCount},
        () => new Array(graph.vertexCount).fill(false)
    );

    for (let edge of graph.edges) {
        insertEdge(a, edge, graph.directed);
    }

    return a;
}

function insertEdge(a: boolean[][], edge: Edge, directed: boolean): void {
    const [x, y] = edge;
    a[x][y] = true;

    if (!directed) {
        insertEdge(a, edge.reverse(), true);
    }
}
